### PR TITLE
Begin prototyping Log API/SDK

### PR DESCRIPTION
**Pull request by tigrannajaryan**
@open-telemetry/go-maintainers This begins implementation of Log API and SDK as defined in the spec https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/README.md, primarily to assess whether the spec is implementable in Go. If we like the code we can keep it and continue working on it to have a complete implementation.

- Most of the the code is copied from trace API/SDK implementation and adopted for logging.
- Added stdoutlog exporter.
- Added sample usage in namedtracer example.
- Tests mostly don't exist for now.
- I did not add a global logger provider to avoid polluting the top-level module with experimental stuff.

`make test` passes successfully. Running namedtracer example will now output logs in addition to spans. Logs will correctly reference the trace and span in the context of which they are emitted.

If we are happy with this approach I will tidy up the PR and will mark it as ready for review. There is no need for detailed line-by-line review yet, please just provide a high-level feedback about the approach, structure of packages, etc.

The next thing I would like to confirm is that this API/SDK is suitable for implementing a handler for https://pkg.go.dev/golang.org/x/exp/slog (in a separate PR, this is already too large).

I can also break this down into smaller PRs if you prefer, but I wanted to show the whole picture for now.

Questions:
- How do we want to make sure this is marked as experimental? Perhaps "x/exp" in package patth names, e.g. go.opentelemetry.io/otel/x/exp/log for API and go.opentelemetry.io/otel/sdk/x/exp/log for SDK? (following convention of experimental packages in Go's standard library).

### CODE REVIEW

**Review by MrAlias — COMMENTED** [review]
I'm not familiar enough with the logging specification to say if this is compliant, but what you have here seems mostly in line with this projects coding standards.

The versioning would be handled by making the `log` and `sdk/log` packages their own modules so they can be tracked with their own pre-1.0 version.

**Comment by tigrannajaryan** [comment]
> I'm not familiar enough with the logging specification to say if this is compliant, but what you have here seems mostly in line with this projects coding standards.
> 
> The versioning would be handled by making the `log` and `sdk/log` packages their own modules so they can be tracked with their own pre-1.0 version.

@MrAlias thanks for reviewing. How do you want to move forward? Do you want me to cleanup this large PR or break it down to smaller ones? I can probably split into 4 PRs:
- Add Log API
- Add Log SDK
- Add stdlogout exporter
- Modify namedtracer example to use logs

**Comment by MrAlias** [comment]
> @MrAlias thanks for reviewing. How do you want to move forward? Do you want me to cleanup this large PR or break it down to smaller ones? I can probably split into 4 PRs:

Yeah, breaking down into some smaller PRs would be ideal. Especially when we consider the tests that will be needed.

I think the SDK could be broken down a bit more given this is where a lot of testing will be added. Maybe:

* Add Log API
* Add Log SDK
  * Providers and processor interfaces
  * Logger
  * Simple processor
  * Batch processor
* Add stdlogout exporter
* Modify namedtracer example to use logs

We usually track this kind of project with a GitHub project. I think that makes sense here to do that as well.

I would also add an issue to the project to include an OTLP exporter as that is the lingua franca of OTel.

**Comment by MrAlias** [comment]
I think the project is a good idea to help scope the work the SIG will need to do to support this. We should then better understand how to prioritize and stage it.

**Comment by tigrannajaryan** [comment]
Now that `slog` added support for Context I wanted to see how it will work with Otel. I think it works nicely.

File: example/namedtracer/foo/foo.go [hunk 1]
```diff
@@ -28,16 +31,23 @@ var (
...
 	defer span.End()
 	span.SetAttributes(lemonsKey.String("five"))
 	span.AddEvent("Sub span event")
 
+	// Log via Otel Logger API
+	logger.Emit(ctx, log.WithAttributes(attribute.String("operation", "suboperation")))
+
+	// Log via slog API. Note this will correctly output Span Context (traceid,spanid) in the final output
+	// if Otel slog Handler is used.
+	slog.InfoCtx(ctx, "Otel logs via slog and with context!", "source", "slog")
```
Comment (tigrannajaryan) [inline example/namedtracer/foo/foo.go:50]: @MrAlias See how this correctly does Otel context-aware logging although it is totally unaware of what Otel is!

**Comment by MadVikingGod** [comment]
Should creating connectors, like slog, logrus, logr, etc., be part of the scope of this?

**Comment by pellared** [comment]
I quickly (~1 hour) looked at the prototype and it looks nice.

I think that `LogRecord` should be `struct` instead of an `interface` to avoid indirect calls and improve the performance. The implementation could be inspired by https://github.com/golang/go/blob/master/src/log/slog/record.go. It would be still compliant with [the OTel Specification](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/).

**Comment by tigrannajaryan** [comment]
@pellared thanks for looking. I mainly mimicked what `trace` package is doing, where Span/TracerProvider are interfaces too. I don't mind any improvements of course if there is significant performance impact, but consistency of the codebase may be a concern too.

**Comment by pellared** [comment]
I am totally aware of it, but in some codebases the logging is a lot more "invasive" than metrics and tracing. From https://youtu.be/tC4Jt3i62ns?si=asGFI7Ni-c7sLPDf&t=168:

> Each request can produce thousands of log messages.

Thus my proposal.

**Pull request closed without merging**